uwnet/stochastic_parameterization/utils.py

```diff
@@ -0,0 +1,200 @@
+import numpy as np
+from functools import lru_cache
+import torch
+from uwnet.tensordict import TensorDict
+import xarray as xr
+
+model_dir = ''
+# model_dir = '/Users/stewart/projects/uwnet/uwnet/stochastic_parameterization/'  # noqa
+model_location = model_dir + 'stochastic_model.pkl'
+default_base_model_location = model_dir + 'full_model/1.pkl'
+default_ds_location = "training.nc"
+# default_ds_location = "uwnet/stochastic_parameterization/training.nc"
+dataset_dt_seconds = 10800
+# default_binning_method = 'precip'
+default_binning_method = 'column_integrated_qt_residuals'
+default_binning_quantiles = (0.06, 0.15, 0.30, 0.70, 0.85, 0.94, 1)
+# default_binning_quantiles = (.1, .3, .7, .9, 1)
+# default_binning_quantiles = (1,)
+# default_binning_quantiles = (.01, .05, .15, .35, .5, .65, .85, .95, .99, 1)
+
+
+class BaseModel(object):
+
+    def __init__(
+            self,
+            model_location,
+            dataset,
+            binning_quantiles=default_binning_quantiles,
+            time_step_days=.125,
+            binning_method=default_binning_method):
+        self.model = torch.load(model_location)
+        self.ds = dataset
+        self.time_step_days = time_step_days
+        self.time_step_seconds = 86400 * time_step_days
+        self.binning_quantiles = binning_quantiles
+        self.binning_method = binning_method
+
+    def get_true_nn_forcing_idx(self, time_idx):
+        start = self.ds.isel(time=time_idx)
+        stop = self.ds.isel(time=time_idx + 1)
+        true_nn_forcing = {}
+        for key in ['QT', 'SLI']:
+            forcing = (start[f'F{key}'].values + stop[f'F{key}'].values) / 2
+            true_nn_forcing[key] = (stop[key].values - start[key].values - (
+                self.time_step_seconds * forcing)) / self.time_step_days
+        return true_nn_forcing
+
+    def predict_for_time_idx(self, time_idx):
+        ds_filtered = self.ds.isel(time=time_idx)
+        to_predict = {}
+        for key_ in self.ds.data_vars:
+            if len(ds_filtered[key_].values.shape) == 2:
+                val = ds_filtered[key_].values[np.newaxis, :, :]
+            else:
+                val = ds_filtered[key_].values.astype(np.float64)
+            to_predict[key_] = torch.from_numpy(val)
+        pred = self.model(TensorDict(to_predict))
+        return {key: pred[key].detach().numpy() for key in pred}
+
+    def get_qt_ratios(self):
+        column_integrated_qt_residuals = np.ones_like(self.ds.Prec.values)
+        moistening_residuals = np.ones_like(self.ds.QT.values)
+        heating_residuals = np.ones_like(self.ds.SLI.values)
+        moistening_preds = np.ones_like(self.ds.QT.values)
+        heating_preds = np.ones_like(self.ds.SLI.values)
+        time_indices = list(range(len(self.ds.time) - 1))
+        idx = 0
+        for time_batch in np.array_split(time_indices, 10):
```

Comment:
This batching in time could be performed automatically using dask, with some small changes to `uwnet.xarray_interface.call_with_xr`.